lib/runtime/wasmer/imports.go

```diff
@@ -244,54 +160,59 @@ func ext_crypto_ed25519_generate_version_1(context unsafe.Pointer, keyTypeID C.i
 
 	if err != nil {
 		logger.Warnf("cannot generate key: %s", err)
-		return 0
+		return []wasmer.Value{wasmer.NewI32(0)}, nil
 	}
 
 	ks, err := runtimeCtx.Keystore.GetKeystore(id)
 	if err != nil {
 		logger.Warnf("error for id 0x%x: %s", id, err)
-		return 0
+		return []wasmer.Value{wasmer.NewI32(0)}, nil
 	}
 
 	err = ks.Insert(kp)
 	if err != nil {
 		logger.Warnf("failed to insert key: %s", err)
-		return 0
+		return []wasmer.Value{wasmer.NewI32(0)}, nil
 	}
 
-	ret, err := toWasmMemorySized(instanceContext, kp.Public().Encode())
+	ret, err := toWasmMemorySized(runtimeCtx, kp.Public().Encode())
 	if err != nil {
 		logger.Warnf("failed to allocate memory: %s", err)
-		return 0
+		return []wasmer.Value{wasmer.NewI32(0)}, nil
+	}
+
+	castedRet, err := safeCastInt32(ret)
+	if err != nil {
+		logger.Errorf("failed to safely cast pointer: %s", err)
+		return []wasmer.Value{wasmer.NewI32(0)}, nil
```

Comment:
The way our code currently operates with regard to runtime functions is that we respond to output variables and validate it on a per-use case basis. So rather than erroring out, we put the burden on the caller to check that the value received is correct.

Here is a conversation about this from another one of my PRs https://github.com/ChainSafe/gossamer/pull/3188#discussion_r1160610879 and I think Kishan's point is valid.

I do, however, wonder if we should migrate away from this and start returning errors from the runtime. I think that would be a pretty large effort and be outside of the scope of this codebase. 

One final note on this specific safe casting case, should we just panic? I know we typically avoid panics, but doing an invalid pointer cast on a pointer used for memory management feels like a potential good reason to me. Let me know your thoughts @timwu20

Reply:
Yup, I would say I agree. I think propagating the errors back to handle is a good way to go.

I think in this case it would be good to pass the error out of the runtime context and then panic. I will open an issue addressing the rest of the runtime error-handling logic

Reply:
Here is the issue: https://github.com/ChainSafe/gossamer/issues/3216